ocaml/xapi/helpers.ml

```diff
@@ -250,6 +250,26 @@ let make_rpc ~__context rpc : Rpc.response =
     else SSL(SSL.make ~use_stunnel_cache:true (), Pool_role.get_master_address(), !Xapi_globs.https_port) in
   XMLRPC_protocol.rpc ~srcstr:"xapi" ~dststr:"xapi" ~transport ~http rpc
 
+let make_timeboxed_rpc ~__context timeout rpc : Rpc.response =
+  let subtask_of = Ref.string_of (Context.get_task_id __context) in
+  Server_helpers.exec_with_new_task "timeboxed_rpc" ~subtask_of:(Context.get_task_id __context) (fun __context ->
+    let open Xmlrpc_client in
+    let http = xmlrpc ~subtask_of ~version:"1.1" "/" in
+    let task_id = Context.get_task_id __context in
+    let cancel () =
```

Comment:
There is significant overlap with `make_rpc`. How about modifying that function instead and adding an optional `timeout` parameter?

Reply:
Perhaps `call_api_functions` ⬇️ can take the `timeout` param as well, and pass it to `make_rpc`, which it calls internally. But then we'd have to make it so that the returned `rpc` function sets the timer itself each time it is called, rather than the make-function doing this.

Reply:
We could probably coalesce them a little. Also, as above, the make_rpc function is only partially applied until it gets the 'rpc' argument, so the timer is indeed not set until you're actually doing an RPC.